### PR TITLE
Clean up wolfSSL_clear() and add some more logging

### PR DESCRIPTION
**Pull request by julek-wolfssl**
Implement WOLFSSL_MSG_EX() which uses XVSNPRINTF to allow for formatted strings to be printed. It uses a 100 byte internal stack buffer to format the log message.

### CODE REVIEW

File: wolfcrypt/src/logging.c [hunk 1]
```diff
@@ -364,6 +364,27 @@ static void wolfssl_log(const int logLevel, const char *const logMessage)
...
+#ifdef __clang__
+/* tell clang argument 1 is format */
+__attribute__((__format__ (__printf__, 1, 0)))
+#endif
+#if !defined(_WIN32) && defined(XVSNPRINTF)
+#include <stdarg.h> /* for var args */
+void WOLFSSL_MSG_EX(const char* fmt, ...)
+{
+    if (loggingEnabled) {
+        char msg[100];
```
Comment (dgarske) [inline wolfcrypt/src/logging.c:376]: Consider making `100` an overridable build-time macro?

File: wolfcrypt/src/logging.c [hunk 2]
```diff
@@ -364,6 +364,27 @@ static void wolfssl_log(const int logLevel, const char *const logMessage)
 }
 
 #ifndef WOLFSSL_DEBUG_ERRORS_ONLY
+#ifdef __clang__
+/* tell clang argument 1 is format */
+__attribute__((__format__ (__printf__, 1, 0)))
+#endif
+#if !defined(_WIN32) && defined(XVSNPRINTF)
```
Comment (dgarske) [inline wolfcrypt/src/logging.c:371]: It looks like types.h always defines `XVSNPRINTF`. Consider having another way to disable? Also if disabled should the format arg still be sent to `wolfssl_log`?

File: src/ssl.c [hunk 3]
```diff
@@ -17989,14 +17995,17 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
...
 
-#ifdef SESSION_CERTS
-        ssl->session->chain.count = 0;
-#endif
 #ifdef KEEP_PEER_CERT
         FreeX509(&ssl->peerCert);
         InitX509(&ssl->peerCert, 0, ssl->heap);
 #endif
 
+        wolfSSL_SESSION_free(ssl->session);
```
Comment (dgarske) [inline src/ssl.c:18003]: Good fix!

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 3 above)
Comment (julek-wolfssl) [reply src/ssl.c]: Actually this should only be done if the session is invalid ie: the handshake didn't complete.

**Review by dgarske — APPROVED** [review]

**Pull request merged**